### PR TITLE
Fix: Corrected file paths when generating CTs, APIs, Controllers, Middlewares, Policies, Services inside Plugins

### PR DESCRIPTION
**Pull request by sunnysonx**
### What does it do?

Fixed file paths for Generator when creating them inside the plugin:
- CTs
- APIs
- Controllers
- Middlewares
- Policies
- Services

### Why is it needed?

Currently, when you generate a plugin through CLI, all `CTs, APIs, controllers, middlewares, policies, services` files are located inside `server` directory: `src/plugins/{pluginName}/server/`,  and that's the correct behavior. 
image

But when you try to generate separately one of the following `CTs, APIs, controllers, middlewares, policies, services` inside a plugin, these are generated directly in the `src/plugins/{pluginName}` directory, instead of having them generated in /server folder. So you end up having two directories. In that case all the  `CTs, APIs, controllers, middlewares, policies, services` that are generated directly inside plugin's directly do not have any effect and are not imported to strapi-server.js
image

### CODE REVIEW

File: packages/generators/generators/lib/plops/content-type.js [hunk 1]
```diff
@@ -80,7 +81,7 @@ module.exports = plop => {
         return Object.assign(object, { [answer.attributeName]: val }, {});
       }, {});
 
-      const filePath = getFilePath(answers.destination);
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/content-type.js:84]: You can, instead, update the `getFilePath` method to add the `/server` suffix to the plugin path:

https://github.com/strapi/strapi/blob/3706033d9fbf994376f31567d3adb659779854fd/packages/generators/generators/lib/plops/utils/get-file-path.js#L9

File: packages/generators/generators/lib/plops/controller.js [hunk 2]
```diff
@@ -18,7 +18,8 @@ module.exports = plop => {
       ...getDestinationPrompts('controller', plop.getDestBasePath()),
     ],
     actions(answers) {
-      const filePath = getFilePath(answers.destination);
+      const resolveFilePath = (isPlugin,filePath) => isPlugin ? `${filePath}/server` : filePath;
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/controller.js:22]: Same as for the CT generator. This should be fixed inside the `getFilePath` method.

File: packages/generators/generators/lib/plops/service.js [hunk 3]
```diff
@@ -16,7 +16,8 @@ module.exports = plop => {
       ...getDestinationPrompts('service', plop.getDestBasePath()),
     ],
     actions(answers) {
-      const filePath = getFilePath(answers.destination);
+      const resolveFilePath = (isPlugin,filePath) => isPlugin ? `${filePath}/server` : filePath;
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/service.js:20]: Same as for the CT generator. This should be fixed inside the `getFilePath` method.

**Review by Elhebert — CHANGES_REQUESTED** [review]
Thanks for the PR 👌

I would suggest updated the path in the `getFilePath` utils instead of adding a new conditional to the generators.

File: packages/generators/generators/lib/plops/content-type.js (hunk 1 above)
Comment (sunnysonx) [reply packages/generators/generators/lib/plops/content-type.js]: done

File: packages/generators/generators/lib/plops/controller.js (hunk 2 above)
Comment (sunnysonx) [reply packages/generators/generators/lib/plops/controller.js]: done

File: packages/generators/generators/lib/plops/service.js (hunk 3 above)
Comment (sunnysonx) [reply packages/generators/generators/lib/plops/service.js]: done

**Comment by sunnysonx** [comment]
@Elhebert 
-Removed/reverted the `/server` path that I have added in previous commits in generators
-Updated the path in the getFilePath utils and added `/server` there
-Refactored middlewares/policies and added the same logic with getFilePath utils (these contained a different logic because these can be created inside root folder)
-Refactored cli prompt answers in get-destinations-prompts to allow 'root' value for middlewares/policies.

Please review

File: packages/generators/generators/lib/plops/policy.js [hunk 4]
```diff
@@ -15,14 +16,7 @@ module.exports = plop => {
...
     actions(answers) {
-      let filePath;
-      if (answers.destination === 'api') {
-        filePath = `api/{{api}}`;
-      } else if (answers.destination === 'plugin') {
-        filePath = `plugins/{{plugin}}/server`;
-      } else {
-        filePath = `./`;
-      }
+      const filePath = getFilePath(answers.destination);
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/policy.js:20]: You can't use `getFilePath` in this case.

The "default" value returned by `getFilePath` is `api/{{ id }}`, where as the policy default value is `./`.
Either you adapt the `getFilePath` method to have an argument to set a default value or you leave the code the way it was.

Both options are valid. I think that keeping the code as is might be the simplest as we don't add complexity to the `getFilePath` function for 2 use cases.

File: packages/generators/generators/lib/plops/middleware.js [hunk 5]
```diff
@@ -15,15 +16,7 @@ module.exports = plop => {
...
-      let filePath;
-      if (answers.destination === 'api') {
-        filePath = `api/{{ api }}`;
-      } else if (answers.destination === 'plugin') {
-        filePath = `plugins/{{ plugin }}/server`;
-      } else {
-        filePath = `./`;
-      }
-
+      const filePath = getFilePath(answers.destination);
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/middleware.js:19]: Same as for the `policy` generator.

File: packages/generators/generators/lib/plops/api.js [hunk 6]
```diff
@@ -45,7 +46,8 @@ module.exports = plop => {
       },
     ],
     actions(answers) {
-      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}/server' : 'api/{{id}}';
+      const answerDestination = answers.isPluginApi && answers.plugin ? 'plugin' : 'api';
+      const filePath = getFilePath(answerDestination);
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/api.js:50]: `getFilePath` isn't compatible with this use case.

The API generator `filePath` wil either be `plugins/ {{ plugin }}/server` or `api/{{ id }}`. But with you changes it will be `plugins/ {{ plugin }}/server` or `api/{{ api }}`.

I would revert this change for simplicity and avoiding adding more code than necessary.

**Review by Elhebert — CHANGES_REQUESTED** [review]
It's a step in the right direction. However you also updated generators to use `getFilePath` when it's not compatible with them.

I added some comments on how to fix them.

**Comment by Elhebert** [comment]
You can disregard my last review. I don't know why GitHub wasn't showing all the changes 🤔

**Review by Elhebert — CHANGES_REQUESTED** [review]
LGTM. One comment about the API generator (see previous review).

**Comment by sunnysonx** [comment]
> LGTM. One comment about the API generator (see previous review).

Hello @Elhebert, can you please make the necessary changes if you will have some available time? I don't have access to my pc right now, and I don't want to modify it from mobile without testing 😞  Thank you

File: packages/generators/generators/lib/plops/api.js [hunk 7]
```diff
@@ -45,7 +46,8 @@ module.exports = plop => {
       },
     ],
     actions(answers) {
-      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}' : 'api/{{id}}';
+      const answerDestination = answers.isPluginApi && answers.plugin ? 'plugin' : 'api';
+      const filePath = getFilePath(answerDestination);
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/api.js:50]:
```suggestion
      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}' : 'api/{{id}}';
```

**Review by Elhebert — COMMENTED** [review]
Hey @sunnysonx,

Sorry for the late response, I've was out of office for a while. Would you be able to do the requested change now or should I take over it? As this is a PR made from your remote I don't think I can make the change directly on it.

I made a suggestion which will allow you to commit the change from GitHub, thus not needing access to your computer.

**Comment by sunnysonx** [comment]
> Hey @sunnysonx,
> 
> Sorry for the late response, I've was out of office for a while. Would you be able to do the requested change now or should I take over it? As this is a PR made from your remote I don't think I can make the change directly on it.
> 
> I made a suggestion which will allow you to commit the change from GitHub, thus not needing access to your computer.

Hello,
Sorry I forgot about this PR. 
Thank you for commit suggestion, I have updated it.

File: packages/generators/generators/lib/plops/api.js [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 const { join } = require('path');
 const fs = require('fs-extra');
+const getFilePath = require('./utils/get-file-path');
```
Comment (Elhebert) [inline packages/generators/generators/lib/plops/api.js:5]: The linter is failing because this import is no longer required.

```suggestion
```

**Review by Elhebert — CHANGES_REQUESTED** [review]

**Review by Elhebert — APPROVED** [review]
LGTM, thanks!
